### PR TITLE
bugfix/resolve-name-aliases

### LINKED ISSUE #65: There are two Dan Strausses (and "No sponsor" person) in Seattle-staging
### Describe the Bug

image
There are two Dan/Daniel Strauss for seattle-staging's `Person` collection.

See https://github.com/CouncilDataProject/cdp-frontend/pull/122#issuecomment-984148009 for discussion.

### Expected Behavior

Only one Dan Strauss should be in DB.

Not sure if this is a backend or scraper issue. We need to investigate what is the cause of the issue. Some possible leads:

- It is the same person just simply has two Legistar entries?
- It could also be that there was a name change in one legistar entry, but our pipeline's "primary keys" for a person think it's a new person (because of different name).
- This could also be that our Seattle static has the wrong name: https://github.com/CouncilDataProject/cdp-scrapers/blob/main/cdp_scrapers/instances/seattle-static.json#L39

cc @dphoria @JacksonMaxfield

### PR DESCRIPTION
**Pull request by dphoria**
### Link to Relevant Issue

This pull request resolves #65 

### Description of Changes

_Include a description of the proposed changes._

We found out the Seattle clerk entered a new entity "Daniel Strauss" into the Legistar DB in place of the correct "Dan Strauss" in a few occasions. They have different primary keys in Legistar. In hindsight this is not too surprising, though not ideal. We should also expect typos, e.g. "Dan Straus".

`file_search_full_form_names(name: str) -> Set[str]`
Search behindthename names file on disk to get full forms of input name.

`query_full_form_names(name: str) -> Set[str]`
Query behindthename API to get full forms of input name.

`name_full_forms(name: str) -> Set[str]`
Call the above methods, in order of preference, to get full forms of input name. The disk file is preferred because API usage is quota-limited. Results are cached to help in performance.

`is_same_person(name1: str, name2: str) -> bool`
Return true if input names are identical or fuzzy-matched.

```Python
from cdp_scrapers.legistar_utils import get_legistar_person
from cdp_scrapers.instances.seattle import SeattleScraper
from cdp_scrapers.scraper_utils import is_same_person

wrong_person = get_legistar_person(client="seattle", person_id=615)
print(wrong_person["PersonFullName"])
# Daniel Strauss

correct_person = get_legistar_person(client="seattle", person_id=664)
print(correct_person["PersonFullName"])
# Dan Strauss

s = SeattleScraper()
correct_cdp_person = s.inject_known_person(s.get_person(correct_person))
fixed_cdp_person = s.inject_known_person(s.get_person(wrong_person))
print(correct_cdp_person == fixed_cdp_person, correct_cdp_person.name, fixed_cdp_person.name)
# True Dan Strauss Dan Strauss

is_same_person("Daniel Strauss", "Dan Strauss") # True
is_same_person("Danny Strauss", "Dan Strauss") # True
is_same_person("Dan Strauss", "Dan Straus") # True
is_same_person("Robert Strauss", "Bob Straus") # True
is_same_person("Dave Strauss", "Dan Strauss") # False
is_same_person("Dick Strauss", "Richard Straus") # True
is_same_person("M. Lorena González", "Lorena Gonzalez") # True
```

### CODE REVIEW

**Comment by dphoria** [comment]
You all have way more experience with fuzzy logic, so please let me know if this is kind of weird and/or over/under-engineered. This is me after reading through web resources for an hour. :sweat_smile:

**Comment by dphoria** [comment]
FYI at the moment most if not all `get_events()` call will return an empty list because seattlechannel.org's certificate has expired, as of 01/15. :rofl: You cannot make this up. i.e. `SeatleScraper` will scrape 0 sessions. I will make some test calls with this PR code after they have resolved that.

**Comment by dphoria** [comment]
Actually, want to change the logic slightly. Converting to draft momentarily.

**Comment by tohuynh** [comment]
A suggestion if you are still using behindthename -- the API version is more stable and plus you wouldn't have to scrape for the results. https://www.behindthename.com/api/help.php (related names)

**Comment by dphoria** [comment]
Nice thank you To. That'll help immensely.

**Comment by dphoria** [comment]
Hm `SeattleScraper.get_events(begin=datetime(2021,9,14), end=datetime(2021,9,15))` still yielded a `Vote` record with `Vote.person = Person(name="Daniel Strauss")`. Arg...

File: cdp_scrapers/scraper_utils.py [hunk 1]
```diff
@@ -66,6 +90,227 @@ def str_simplified(input_str: str) -> str:
...
+    if len(name) == 0 or len(query_name) == 0:
+        return False
+
+    # can't always use the first substring as first name
+    # e.g. first_initial middle_name last_name
+    name_parts = name.split()
+    query_name_parts = query_name.split()
+
+    for i in range(len(name_parts)):
+        if len(name_parts[i]) <= 1 and len(name_parts) <= 2:
```
Comment (tohuynh) [inline cdp_scrapers/scraper_utils.py:274]: `for i, name_part in enumerate(name_parts):` is more pythonic

**Review by tohuynh — COMMENTED** [review]
When you fixed the bug, it would be good to add a few unit test cases for `is_same_person`

**Comment by tohuynh** [comment]
Eh there's gotta be a free name matching API out there we can use? :(

File: cdp_scrapers/scraper_utils.py (hunk 1 above)
Comment (dphoria) [reply cdp_scrapers/scraper_utils.py:274]: Ah, thank you!

**Comment by dphoria** [comment]
> Eh there's gotta be a free name matching API out there we can use? :(

That was my first thought. I thought for sure. But all I found were basically tutorials doing some of what I did and then much much more. I will look again. I completely agree we don't want to have to reinvent the wheel here.

**Comment by tohuynh** [comment]
> > Eh there's gotta be a free name matching API out there we can use? :(
> 
> That was my first thought. I thought for sure. But all I found were basically tutorials doing some of what I did and then much much more. I will look again. I completely agree we don't want to have to reinvent the wheel here.

I was only able to find this one: https://www.nameapi.org/en/developer/manuals/rest-web-services/53/web-services/name-matcher/

But the demo doesn't always work for me. 

Happy to have this PR as is, fix our Dan/Daniel bug.

**Comment by JacksonMaxfield** [comment]
I will try to review this tonight sorry for the delay

Early thought:
1. we shouldn't store your own API key in the repo. That is probably already scraped and stored somewhere else :joy: 
2. I _really_ don't want to use an external service unless we _need_ to.
3. for now, why can't we do it on a per-instance basis. Much like how we have per-scraper minutes item filtering, we should have per-scraper names matching. This is just generally "safer" imo because what if we have a councilmember bob boberson one year and then 6 years later we have robert boberson, a different person. this fuzzy matching will result in a match if i am following it correctly. I would argue that, while fuzzy matching is _cool_ it is unsafe in this usage.

**Comment by dphoria** [comment]
>     3. for now, why can't we do it on a per-instance basis. Much like how we have per-scraper minutes item filtering, we should have per-scraper names matching. This is just generally "safer" imo because what if we have a councilmember bob boberson one year and then 6 years later we have robert boberson, a different person. this fuzzy matching will result in a match if i am following it correctly. I would argue that, while fuzzy matching is _cool_ it is unsafe in this usage.

First of all, yes you are correct in the bob boberson v. robert boberson example (lovely name LOL). And I do think there is a case to made here, To (me being the world's most undecisive person). Especially in the spirit of "keep it stupid simple."

i.e. Seattle define explicitly, "Daniel Strauss is Dan Strauss."

**Comment by tohuynh** [comment]
@dphoria Does "Daniel Strauss" have any "Councilmember" roles?

**Comment by dphoria** [comment]
I queried Seattle Legistar API directly for OfficeRecords/Roles for that person and I got just 1 entry, as a "member."

```
{
    "OfficeRecordId":1127,
    "OfficeRecordGuid":"851A17AC-8D60-4750-9DB8-0EB571B8FEEB",
    "OfficeRecordLastModifiedUtc":"2020-01-05T23:11:50.167",
    "OfficeRecordRowVersion":"AAAAAACbkbA=",
    "OfficeRecordFirstName":"Daniel",
    "OfficeRecordLastName":"Strauss",
    "OfficeRecordEmail":"",
    "OfficeRecordFullName":"Daniel Strauss",
    "OfficeRecordStartDate":"2020-01-06T00:00:00",
    "OfficeRecordEndDate":"2021-12-31T00:00:00",
    "OfficeRecordSort":5,
    "OfficeRecordPersonId":615,
    "OfficeRecordBodyId":248,
    "OfficeRecordBodyName":"Governance and Education Committee",
    "OfficeRecordTitle":"Member",
    "OfficeRecordVoteDivider":1.0,
    "OfficeRecordExtendFlag":0,
    "OfficeRecordMemberTypeId":3,
    "OfficeRecordMemberType":"Member",
    "OfficeRecordSupportNameId":null,
    "OfficeRecordSupportFullName":null
}
```

**Comment by JacksonMaxfield** [comment]
> @dphoria Does "Daniel Strauss" have any "Councilmember" roles/officerecords?

And in our db he has none:
```ipython
In [1]: from cdp_backend.database import models as db_models
   ...: import fireo
   ...: from google.auth.credentials import AnonymousCredentials
   ...: from google.cloud.firestore import Client
   ...: 
   ...: # Connect to the database
   ...: fireo.connection(client=Client(
   ...:     project="cdp-seattle-staging-dbengvtn",
   ...:     credentials=AnonymousCredentials()
   ...: ))

In [2]: daniel = list(db_models.Person.collection.filter("name", "==", "Daniel Strauss").fetch())[0]

In [3]: roles = list(db_models.Role.collection.filter("person_ref", "==", daniel.key).fetch())

In [4]: roles
Out[4]: []
```

**Comment by dphoria** [comment]
I'll do pull an event to get Daniel Strauss as CDP `Person`, but pretty sure it will reflect that one record. One moment...

**Comment by tohuynh** [comment]
We would get away from this Dan/Daniel bug by just displaying people on the `/people` only if they have "Councilmember" roles.

But this name bug could reappear if the situation is different -- somehow the other "name" also has legit data. Then we would really need `is_same_person`

**Comment by tohuynh** [comment]
> And in our db he(Daniel) has none:

Shouldn't Daniel have at least one role: Member of Governance and Education Committee? (there is an OfficeRecord, according to Sung).

**Comment by JacksonMaxfield** [comment]
> > And in our db he(Daniel) has none:
> 
> Shouldn't Daniel have at least one role: Member of Governance and Education Committee? (there is an OfficeRecord, according to Sung).

We cleared out all of seattle staging after the hashing bugfix. Since legistar is now returning "correct" data for all of the events we have reprocessed, we haven't received the bad "Daniel Strauss" again since then!

**Comment by tohuynh** [comment]
Then "Daniel Strauss" shouldn't be in `Person` too right?

**Comment by JacksonMaxfield** [comment]
> Then "Daniel Strauss" shouldn't be in `Person` too right?

Correct. Now I am confused... Will check voting history or something.

**Comment by tohuynh** [comment]
Could you check MatterSponsor too?

**Comment by JacksonMaxfield** [comment]
> Could you check MatterSponsor too?

No matters sponsored

**Comment by dphoria** [comment]
> > > And in our db he(Daniel) has none:
> > 
> > 
> > Shouldn't Daniel have at least one role: Member of Governance and Education Committee? (there is an OfficeRecord, according to Sung).
> 
> We cleared out all of seattle staging after the hashing bugfix. Since legistar is now returning "correct" data for all of the events we have reprocessed, we haven't received the bad "Daniel Strauss" again since then!

Right, all depends on if we got at least one `EventIngestionModel` with that person as a sponsor or a voter.

**Comment by dphoria** [comment]
Given recent discussions, we will address the issue #65 in a completely approach.

**Pull request closed without merging**